driver-core/src/main/java/com/datastax/driver/core/PagingState.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * PagingState comes with a certified hash verification which allows to certify the validity
+ * of the PagingState buffer. Hence, using a saved PagingState with another Statement than the
+ * one it was made with, will fail the verification and the saved PagingState won't be
+ * usable. Also, if the PagingState content buffer (or the String the PagingState was converted to)
+ * is modified , the verification will fail, too. The content of the PagingState should not be modified
+ * at any time.
+ *
+ */
+
+public class PagingState implements Serializable{
+
+    private static final long serialVersionUID = 7526471155622776147L;
+    private ByteBuffer pagingState;
+    private ByteBuffer hashDigest;
+
+    private PagingState(ByteBuffer pagingState, Statement statement) {
+        this.pagingState = pagingState;
+        if (statement != null)
+            this.hashDigest = ByteBuffer.wrap(generatePagingStateHash(statement));
+    }
+
+    public PagingState(ByteBuffer pagingState, ByteBuffer hashDigest){
+        this.pagingState = pagingState;
+        this.hashDigest = hashDigest;
+    }
+
+    public static PagingState wrap(ByteBuffer pagingState, Statement statement) {
+        return new PagingState(pagingState, statement);
+    }
+
+    public ByteBuffer getPagingState() {
+        return this.pagingState;
+    }
+
+
+    private byte[] generatePagingStateHash(Statement statement) {
+        byte[] digest = null;
+
+        assert !(statement instanceof BatchStatement);
+
+        ByteBuffer[] values;
+        MessageDigest md;
+        try {
+            md = MessageDigest.getInstance("MD5");
+            if (statement instanceof BoundStatement){
+                BoundStatement bs = ((BoundStatement)statement);
+                md.update(bs.preparedStatement().getQueryString().getBytes());
+                values = bs.values;
+            }
+            else if (statement instanceof RegularStatement){
+                RegularStatement rs = (RegularStatement)statement;
+                md.update(rs.getQueryString().getBytes());
+                values = rs.getValues();
+            }
+            else {
+                throw new InvalidQueryException("Statement type invalid for generating paging check, "
+                    + "you must use a SimpleStatement(/Regular Statement) or a PreparedStatement(/Bound Statement). "
+                    + "Batch Statement are not made for the usage of Cassandra pager.");
+            }
+            if (this.pagingState != null){
+                if (values != null){
+                    for(int i = 0 ; i < values.length ; i++){
+                        md.update(Bytes.getArray(values[i]));
+                    }
+                }
+                ByteBuffer psCopy = this.getPagingState().duplicate();
+                byte[] pagingState = new byte[psCopy.remaining()];
+                psCopy.get(pagingState);
+                md.update(pagingState);
```

Comment:
`update` can also take a byte buffer, so you could do:

    if (values != null) {
        for (ByteBuffer value : values)
            md.update(value.duplicate());
    }
    md.update(this.getPagingState().duplicate());